File: __old-version/apis/server/test/alexa_test.js
/* */

const chai = require('chai');
const chaiHttp = require('chai-http');
const chaiXml = require('chai-xml');

const expect = require('chai').expect;

chai.use(chaiHttp);
chai.use(chaiXml);

const url = 'https://api.codetabs.com';
// const url = 'http://localhost:3000'

describe('ALEXA TEST ', function () {
  before(function (done) {
    done();
  });
  it('not domain (empty)', function (done) {
    chai.request(url)
      .get('/alexa/get/')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('error', 'Domain is empty');
        done();
      });
  });
  it('not a valid domain name', function (done) {
    chai.request(url)
      .get('/alexa/get/code%%tabs.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        // expect(res).to.be.json
        // expect(res.body).to.be.an('object')
        // expect(res.body).to.have.property('error', 'not a valid domain name')
        done();
      });
  });
  it('domain in query string', function (done) {
    chai.request(url)
      .get('/alexa/get')
      .query({d: 'google.com'})
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body.rank).to.be.oneOf(['1', '2', '3', '4', '5']);
        expect(res.body).to.have.property('domain', 'google.com');
        expect(res.body).to.have.property('rank');
        done();
      });
  });
  it('domain in route parameters ', function (done) {
    chai.request(url)
      .get('/alexa/get/google.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body.rank).to.be.oneOf(['1', '2', '3', '4', '5']);
        expect(res.body).to.have.property('domain', 'google.com');
        expect(res.body).to.have.property('rank');
        done();
      });
  });
  it('www.github.com', function (done) {
    chai.request(url)
      .get('/alexa/get/www.github.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'github.com');
        done();
      });
  });
  it('github.com', function (done) {
    chai.request(url)
      .get('/alexa/get/github.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'github.com');
        done();
      });
  });
  it('valid domain not in alexa top 1m', function (done) {
    const text1 = 'not-top-domain.com not in alexa top 1 million';
    chai.request(url)
      .get('/alexa/get/not-top-domain.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('error', text1);
        done();
      });
  });
  it('awwwards.com ,www in the middle of the domain', function (done) {
    chai.request(url)
      .get('/alexa/get/awwwards.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'awwwards.com');
        done();
      });
  });
  it('progresswww.nl ,www at the end of the domain', function (done) {
    chai.request(url)
      .get('/alexa/get/progresswww.nl')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'progresswww.nl');
        done();
      });
  });
  it('www.gov.uk ,www as subdomain', function (done) {
    chai.request(url)
      .get('/alexa/get/www.gov.uk')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'www.gov.uk');
        done();
      });
  });
  it('www.nic.in ,www as subdomain', function (done) {
    chai.request(url)
      .get('/alexa/get/www.nic.in')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'www.nic.in');
        done();
      });
  });
  it('bih.nic.in ,not www subdomain in a multiple results', function (done) {
    chai.request(url)
      .get('/alexa/get/bih.nic.in')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'bih.nic.in');
        done();
      });
  });
  it('nic.in , multiple results include the name ', function (done) {
    const text2 = 'Multiple results include nic.in .Must be more specific';
    chai.request(url)
      .get('/alexa/get/nic.in')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('error', text2);
        done();
      });
  });
});

File: alexa/alexa_test.go
/* */

package alexa

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestAlexaApi(t *testing.T) {

	const (
		validFormat = "Bad request, valid format is 'api.codetabs.com/v1/{service}?{param}=value' .Please read our docs at https://codetabs.com"
	)

	a := NewAlexa(true)

	type alexaTestOutput struct {
		Domain     string `json:"domain"`
		Rank       int    `json:"rank"`
		StatusCode int    `json:"statusCode"`
		Error      string `json:"Error,omitempty"`
	}

	var alexaTests = []struct {
		it         string
		endpoint   string
		errorText  string
		statusCode int
	}{
		{"not domain(empty)", "/v1/alexa", validFormat, 400},
		{"not domain(empty)", "/v1/alexa/", validFormat, 400},
		{"not valid format", "/v1/alexa//", validFormat, 400},
		{"not valid format", "/v1/alexa/&%$/wrong", `parse "/v1/alexa/&%$/wrong": invalid URL escape "%$/"`, 400},
		{"not valid format", "/v1/alexa/web?=codetabs.com", validFormat, 400},
		{"not valid format", "/v1/alexa?we=codetabs.com", validFormat, 400},
		{"not valid format", "/v1/alexa/too/much/parameters", validFormat, 400},
		{"not a valid domain name", "/v1/alexa?web=code%%tabs.com", validFormat, 400},
		{"domain in routes parameters", "/v1/alexa?web=google.com", "", 200},
		{"www.github.com", "/v1/alexa?web=www.github.com", "", 200},
		{"github.com", "/v1/alexa?web=github.com", "", 200},
		{"valid domain not in alexa top 1m", "/v1/alexa?web=not-top-domain.com", "not-top-domain.com not in alexa top 1 million", 400},
		{"awwwards.com ,www in the middle of the domain", "/v1/alexa?web=awwwards.com", "", 200},
		{"progresswww.nl ,www at the end of the domain", "/v1/alexa?web=progresswww.nl", "", 200},
		{"www.gov.uk ,www as subdomain", "/v1/alexa?web=www.gov.uk", "", 200},
		{"gov.uk ,multiple results", "/v1/alexa?web=gov.uk", "", 200},
		{"www.nic.in ,www as subdomain", "/v1/alexa?web=www.nic.in", "", 200},
		{"bih.nic.in ,not www subdomain in multiple results", "/v1/alexa?web=bih.nic.in", "", 200},
		{"nic.in ,multiple results", "/v1/alexa?web=nic.in", "", 200},
	}

	for _, test := range alexaTests {
		var to = alexaTestOutput{}
		pass := true
		req, err := http.NewRequest("GET", test.endpoint, nil)
		if err != nil {
			if err.Error() != test.errorText {
				t.Errorf("Error Request %s\n", err.Error())
			} else {
				pass = false
			}
		}
		if pass {
			rr := httptest.NewRecorder()
			handler := http.HandlerFunc(a.Router)

			handler.ServeHTTP(rr, req)
			if rr.Code != test.statusCode {
				t.Errorf("%s got %v want %v\n", test.endpoint, rr.Code, test.statusCode)
			}
			_ = json.Unmarshal(rr.Body.Bytes(), &to)
			if to.Error != test.errorText {
				t.Errorf("%s got %v want %v\n", test.endpoint, to.Error, test.errorText)
			}
		}
		fmt.Printf("Test OK...%s\n", test.endpoint)
	}
}
